### PR TITLE
Introduce FilterService as abstraction for creating images

### PR DESCRIPTION
**Pull request by rpkamp**
| Q | A
| --- | ---
| Branch? | 2.0
| Bug fix? | no
| New feature? | no
| BC breaks? | yes
| Deprecations? | no
| Tests pass? | yes
| Fixed tickets |
| License | MIT
| Doc PR |

This PR splits all the image manipulation from the ImagineController and ResolveCacheCommand classes to a new FilterService class, which was on the roadmap for 2.x:

> Move Controller Business Logic into Service (Allow for use of service without Request)

The advantages of this are:
- Controller is only concerned with HTTP infrastructure
- Command is only concerned with CLI I/O
- Controller and command are easier to read, more highlevel
- Functionality is easier to use and extend in custom controllers/command from third parties
- FilterManager#applyFilter is only called in one spot throughout the entire codebase, making it easier to allow for multiple filter implementations later on

This PR is missing tests for the FilterService class, but I intend to add those once everyone agrees on a general direction of this PR. I did test the functionality in the browser (by using it in a separate test project) and that works as expected.

**Discussion**
In the FilterService the method `createFilteredImage` is closely tied to `getUrlOfFilteredImage`, but the first is to create the image and the latter is to retrieve the URL of the created image. Same thing goes for `createFilteredImageWithRuntimeFilters` and `getUrlOfFilteredImageWithRuntimeFilters`.
I split these out because of the CQS principle, but have now introduced a temporal coupling between the two. A more pure CQS approach would be to generate a URL first and then pass it to `createFilteredImage` as the desired URL for the image. However, it would be a mayor overhaul at this point to accomplish that, if indeed it is at all possible.

An alternative could be to let `createFilteredImage` return the URL of the resulting image (whether is was obtained from cache or not) in kind of find-or-create pattern way. This breaks CQS, but might be easier to understand and is less verbose in usage because one would only need one call instead of two to create a filtered image and obtain its URL.

### CODE REVIEW

File: Service/FilterService.php [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        return $this->cacheManager->resolve($runtimePath, $filter, $resolver);
+    }
+
+    /**
+     * @param string $path
+     * @param string $filter
+     * @param array|null $runtimeFilters
+     * @return \Liip\ImagineBundle\Binary\BinaryInterface
+     * @throws \Liip\ImagineBundle\Exception\Imagine\Filter\NonExistingFilterException
+     */
```
Comment (robfrawley) [inline Service/FilterService.php]: All doc blocks need their CS fixes such that they are ordered `param`, `throws`, `return`, each type is separated by a two newlines, and have their `param` variables aligned. We also use short class names, even if they aren't used in the current file a and require an additional `use` statement.
```php
/**
 * @param string     $path
 * @param string     $filter
 * @param array|null $runtimeFilters
 *
 * @throws \Liip\ImagineBundle\Exception\Imagine\Filter\NonExistingFilterException
 *
 * @return \Liip\ImagineBundle\Binary\BinaryInterface
 */
```

Comment (rpkamp) [inline Service/FilterService.php]: I'm confused regarding using short class names, as you said the complete opposite in this comment: https://github.com/liip/LiipImagineBundle/pull/908/files/f9cb8243475a1e492cb77ac5e639d306381fe9a8#diff-3edb79e074a2b20890b43396b316b5d3L60

Comment (rpkamp) [inline Service/FilterService.php]: Or is the short class name only used for `@param` but not for `@throws` and `@return`?

Comment (robfrawley) [inline Service/FilterService.php]: My bad, regarding the referenced comment, no I thought that was a `@covers` which is where we've been adding leading backslashes. That, too, should be fixed, I just skimmed over what it was too fast.

Comment (robfrawley) [inline Service/FilterService.php]: We have a `php-cs-fixer` `require-dev` requirement as well as the associated style settings; I would just run that, which will take care of everything but the FQCNs.

Comment (robfrawley) [inline Service/FilterService.php]: Erm, looks like our fixer rules are out of date and most of them are skipped, so that isn't a sure-fire bet either...

Comment (rpkamp) [inline Service/FilterService.php]: Is this what you mean? https://github.com/liip/LiipImagineBundle/pull/922/commits/4bce009f73c55659143d1c731709aed78b435e19

Comment (robfrawley) [inline Service/FilterService.php]: Short of the FQCNs that still exist, yes.

Comment (robfrawley) [inline Service/FilterService.php]: These instances *should* be the only examples of FQCNs: https://github.com/liip/LiipImagineBundle/blob/1.0/Tests/Binary/Loader/FileSystemLoaderTest.php#L22

Comment (rpkamp) [inline Service/FilterService.php]: Ah yes, that makes sense. Thanks.

**Comment by robfrawley** [comment]
The RST documentation (which is what is pushed to [Symfony's online documentation website](http://symfony.com/doc/current/bundles/LiipImagineBundle/index.html)) requires changes and additional pertinent to this change, as well.

**Comment by rpkamp** [comment]
I've merged this PR with master so I could use short array syntax again :tada:

**Comment by Koc** [comment]
Lets wait until #926 will be merged into 2.0

**Comment by Koc** [comment]
Also closes #806

**Comment by rpkamp** [comment]
Not sure what's going on with https://github.com/liip/LiipImagineBundle/pull/926 since it's closed but not merged?

**Comment by robfrawley** [comment]
The `1.x` branch will be merged with `2.0` once we release `1.8.0` (see #928).

**Comment by Koc** [comment]
You need apply your service also here https://github.com/liip/LiipImagineBundle/blob/2.0/Async/ResolveCacheProcessor.php

**Comment by robfrawley** [comment]
All branches up to date.

**Comment by rpkamp** [comment]
I've updated the `ResolveCacheProcessor` to make use of the new `FilterService`.

**Comment by rpkamp** [comment]
Does anyone have any thoughts on the CQS vs Temporal Coupling discussion in my initial PR message? 

The more I think about it, the more I think I'd rather have a CQS violation than temporal coupling, so instead of one method to query what the URL would be and a command to create the image, create one find-or-create method that creates the image if it doesn't exist and then return the URL, or if the image already existed just return the URL.
The reason I like it better is because it easier to understand and harder to get wrong for people starting out with this bundle.

**Comment by robfrawley** [comment]
@rpkamp 

> Let createFilteredImage return the URL of the resulting image (whether is was obtained from cache or not) in kind of find-or-create pattern way. This breaks CQS, but might be easier to understand and is less verbose.

Yes, I absolutely prefer this approach.

**Comment by rpkamp** [comment]
@robfrawley 

> Yes, I absolutely prefer this approach.

Done. This approach makes more sense for this project than the previous approach :smiley:

**Comment by rpkamp** [comment]
Could I get a review on this one please?

**Comment by Koc** [comment]
Nice PR, can anybody review?

**Comment by makasim** [comment]
LGTM

**Comment by rpkamp** [comment]
This PR has been open for a long time (since May of this year).
Anything I can do to advance the merging of this PR?

**Comment by makasim** [comment]
@rpkamp  Thank you for the contribution.

**Pull request merged**